Show an error if creating the upload archive fails

chrismathis opened the pull request:
e.g if there are duplicate jars.
In my case feature and plug-in had the same name

ctron commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/UploaderV3.java:
```diff
@@ -104,6 +104,7 @@ public void upload ( final File file, final String filename ) throws IOException
         }
         catch ( final IOException e )
         {
+            e.printStackTrace ( this.listener.error ( "Failed to create upload archive" ) );
```
Could you write that to some logger instead?

nfalco79 replied:
@chrismathis  `this.listener` is the jenkins logger to print to the build console so `this.listener.error` is enough

EDIT:
maybe with also the cause `this.listener.error ( "Failed t......because: " e.getMessage() )`

chrismathis replied:
The e.printStackTrace adds the message form the Exception to the log (of this.logger). Doesn't it?
That's copied form Line 177 or Line 221.

Should I do 
`this.listener.error ( "Failed to create upload archive: " + e.getMessage () );`
instead?

nfalco79 replied:
The `e.printStackTrace` prints to the Standard Error that means that it is logged to the jenkins log and is not visible in the jenkins build console. Unless you login to the jenkins master node and find the (system or everywhere you redirects output) logs you never see it. In general in Java application usually the `e.printStackTrace` use is discouraged infavor or any logger library (because could be configured the output)

`this.listener` in this case is the TaskListener that is a Jenkins wrapper for the log to print to the jenkins console log. So if your goal is advise the user why the build is failed than `this.listener` is the one that does your case. Maybe instead just notice "fail to create upload" if you post also the original cause message can be of more help in understanding what the problem is.

ctron replied:
Sorry, I overlooked that. You actually are sending the output to the listener.

My mistake!

chrismathis replied:
@nfalco79 
In this case this.listener.error() provides a PrintWriter to append stuff (the StackTrace) to the console output. (Just retested)

nfalco79 replied:
ok

chrismathis commented:
After some additional experiments I have to say I like logging only the exception message better.
Because the method is called for each file uploaded you might end up with lots of similar stack traces otherwise.

nfalco79 commented:
Could you provide a single squash commit?

ctron commented:
@chrismathis sorry for the confusion I caused. Please revert to the original way of logging this. I was under the impression that this would be logged to stdout.

chrismathis commented:
@ctron the latest change is motivated by the fact that the method is called for each file (which might be many).
This might lead the very many similar Stacktraces in the console.
I recognized that when throwing a new Exception in the upload try-catch to test out the log entries.

Do you prefer the (potentially) many Stacktraces over single line messages?

ctron commented:
I prefer not the have any errors 😁 

I am not sure about this. What you could do is to actually show the full stack trace when the "failed" flag is still false and the compressed message afterwards. That way you would have the full stack trace, but only once.

What do you think?

chrismathis commented:
That sounds good - i'll do that.

nfalco79 commented:
At this point, why do not you check the failed flag at the beginning of the method instead of creating the entries since the work that would be lost?
So I would expect:
```
Failed to upload file1 to archive to upload: Space left on device
file2 was skipped due previous error (<- or log nothing)
file3 was skipped due previous error (<- or log nothing)
...
```

ctron commented:
So you mean not to even try? Sounds reasonable to me.

chrismathis commented:
@nfalco79 you are right. It does not make much sense to write the files to the archive when complete() (which does the actual upload) will do nothing if this.failed.

chrismathis commented:
Should IOExceptions actually be ignored if failsAsUpload is false?
That does not work for V2 at the moment (at least in pipeline jobs). They fail the build.

nfalco79 commented:
I think the UploadV2 simply miss to handle IOException.
I think it should be always catch and log it (but I prefer without whole tracktrace in the build log) Only @ctron could say if the original idea was continue the upload of the following files or stop it. V3 acts like a transaction where all files are uploaded or none, V2 no.
The Upload implementation should be aware of failure strategy. Is the recorder that based on user choice do the work

nfalco79 commented:
I have to correct my previous comment (In these days I'm writing aware from home and PC, so pardon my errors). The interface Uploader have the method upload that returns void and declare to throws IOException. So in the implementation if any IO error occur could throwed. It is the callable UploadFiles that should catch and handle the exception (and maybe log it).
So about the  V3 I can say that the original  implementation simply miss a `throw e` after set the failed flag.

ctron commented:
No worries ;-)

To be honest, I have. I idea what I had in mind :-D

I guess throwing it the same way would be a reasonable way to do it.

chrismathis commented:
I've refactored the Uploaders to throw Exceptions instead of setting a flag. The IOExceptions are caught in UploadFiles and returned by invoke. If they were thrown Jenkins would add: remote file operation failed: " + remote + " at " + channel + ": ".

Reading fields of FileCallables does not seem to work if they are executed on a slave. They do not seem to get serialized back. So isFailed and isEmptyArchive always returned false.

Jobs will now only fail if failsAsUpload is true or allowEmptyArchive is false when the respective errors occur. (the empty archive behavior was inverted)

nfalco79 commented:
The FileCallable serialise back just only the returned value and not the entire instance.
Since the UploadFiles must collect the artifacts to construct the BuildData it should return Map<String, String>. Now map was not filled (there is a comment to remember to fix it).

Personally I do not like that exception design. Do you make in that manner becuase the any raised IOException is not catched in the buildstep? Note that `workspace.act` wraps the IOException into other IOException. Have a look to the hudson.tasks.ArtifactArchiver (to which this class I think has been inspired) on how to handles IOExceptions.

As I suggest to Jens this class should be refactored to be working correctly on slave node.

nfalco79 commented:
The other goal of refactor of the DroneRender is enable this class to be more testable.

chrismathis commented:
@nfalco79 I did it that way because `workspace.act` wraps the IOException and adds it's own error message. That would hide the Uploader exception messages.

The `ArtifactArchiver`  returns a Map<String,String> to feed it to ArtifactManager. Do you plan to return a Map<String,String> to feed it to something else that does the upload from the `perform` method? Will that be executed on the correct machine (master/slave)?
**Edit**: found the usage of the map for run.addAction. Although I could not find where the information from the map adds something to Jenkins UI

If the IOExceptions are not caught at all, they fail the build. I think if failsAsUpload is false the build should not fail if there are IOExceptions with the packagedrone server.

chrismathis commented:
We could add a UploadResult class to contain the Map<String,String> for the action and the Exception or problem flags (`isEmptyUpload`, `uploadFailed`)

nfalco79 commented:
@chrismathis you can unwrap the IOException until you get your UploaderException (extended by Empty....Exception and FailedUpload...Excption), otherwise other kind of issue was happens.

Anyway the UploadResult solution sound good

chrismathis committed:
Improve on error reporting and job fail conditions

*Refactor Uploaders to throw exceptions instead of setting a flag
*Fail only if failsAsUpload is true
*Fail on empty archive if allowEmptyArchive is false (not true)
*Reading fields of a Callable does not work if it is run on a slave as
they are
not serialized back -> return UploaderResult

chrismathis commented:
Added UploaderResult that reports back build status  and artifact map back to master.

nfalco79 commented on package-drone/.settings/org.eclipse.jdt.ui.prefs:
```diff
@@ -28,7 +28,7 @@ sp_cleanup.make_type_abstract_if_missing_method=false
 sp_cleanup.never_use_blocks=false
 sp_cleanup.never_use_parentheses_in_expressions=true
-sp_cleanup.on_save_use_additional_actions=true
+sp_cleanup.on_save_use_additional_actions=false
```
maybe changed for mistake

chrismathis replied:
Actually not by mistake. "Remove trailing whitespace" and "Remove redundant type arguments" kept reformatting unrelated code.
-> Re-enabled and changed the settings for whitespace and type arguments.

nfalco79 replied:
@ctron has setup the formatter according with his other projects so if you want change formatter only him can decide

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/UploaderResult.java:
```diff
@@ -0,0 +1,38 @@
+import java.io.Serializable;
+import java.util.Map;
+
+public class UploaderResult implements Serializable
```
Add missing javadoc

nfalco79 replied:
This is a data transport bean, use the JavaBeans naming conventions.

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/UploaderResult.java:
```diff
@@ -0,0 +1,38 @@
+public class UploaderResult implements Serializable
+{
+    private static final long serialVersionUID = -3089286880912224513L;
+    private Map<String,String> artifacts;
```
initialise this map in the contructor

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/UploaderResult.java:
```diff
@@ -0,0 +1,38 @@
+        return artifacts;
+    }
+    
+    public void addArtifacts(Map<String, String> artifacts){
```
if the method is addArtifats I expect it does no set the entire map but add to existings.

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/UploaderResult.java:
```diff
@@ -0,0 +1,38 @@
+        return isEmptyUpload;
+    }
+    
+    public void setEmptyUpload(boolean isEmpty) {
```
* setEmptyUpload
* isEmptyUpload
* isEmpty
JavaBeans naming conventions

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/UploaderResult.java:
```diff
@@ -0,0 +1,38 @@
+    
+    boolean failed = false;
+    
+    public Map<String,String> getUploadedArtifacts(){
```
if `addArtifats` then `getArtifacts` and I expect retuns an unmodified map (ensure that no one could modify internal data)

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/UploaderV2.java:
```diff
@@ -43,8 +43,8 @@
 
-    private boolean failed;
-
+    private final Map<String, String> uploadedArtifacts = new HashMap<> ();
```
Move to abstract class

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/UploaderV2.java:
```diff
@@ -126,24 +126,25 @@ public void upload ( final File file, final String filename ) throws IOException
-        this.failed = true;
-
         final String message = makeString ( response.getEntity () );
+        String errorMessage = String.format (  "Failed to upload %s: %s %s = %s", fileName, response.getStatusLine ().getStatusCode (), response.getStatusLine ().getReasonPhrase (), message  );
```
externlise this to be i18n

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/UploaderV3.java:
```diff
@@ -148,61 +142,51 @@ public boolean complete ()
                         case 404:
-                            this.listener.error ( "Failed to find upload endpoint V3. This could mean that you configured a wrong server URL or that the server does not support the Upload V3. You will need a version 0.14+ of Eclipse Package Drone. It could also mean that you did use wrong credentials." );
-                            return false;
+                        	throw new IOException ("Failed to find upload endpoint V3. This could mean that you configured a wrong server URL or that the server does not support the Upload V3. You will need a version 0.14+ of Eclipse Package Drone. It could also mean that you did use wrong credentials.");
```
externalise to be i18n

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/UploaderV3.java:
```diff
@@ -267,6 +253,15 @@ public ArtifactInformation getArtifactInformation ()
     }
+    
+    private Map<String, String> createArtifactsMap(final UploadResult result){
+        Map<String, String> artifacts = new HashMap<> ();
```
you can initialise the size of the map based on `result.getCreatedArtifacts ().size()`

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/DroneRecorder.java:
```diff
@@ -341,45 +339,20 @@ public void perform ( final Run<?, ?> run, final FilePath workspace, final Launc
-            }
+      
+        UploaderResult result = workspace.act ( uploader );
+        if ((result.hasFailed () && failsAsUpload) || ( result.isEmptyUpload && ! this.allowEmptyArchive)) {
```
Restore the logic to log when no artifacts there because you could change this logic based on `allowEmptyArchive` value in the future and Uploader does not keep user choice.
The Uploader can not performs all tasks, its task is only upload the requested files.

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/DroneRecorder.java:
```diff
@@ -341,45 +339,20 @@ public void perform ( final Run<?, ?> run, final FilePath workspace, final Launc
-        run.addAction ( new BuildData ( this.serverUrl, this.channel, uploader.artifacts ) );
-    }
-
-    private void logWarning ( TaskListener listener, String message )
```
this method is useful to log warning message without wrote the same prefix

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/DroneRecorder.java:
```diff
@@ -445,33 +402,23 @@ public boolean isEmptyArchive ()
-                this.emptyArchive = true;
-                // nothing to upload
+                uploadResult.setEmptyUpload ( true );
+                this.listener.error(Messages.DroneRecorder_noMatchFound (this.includes)); // nothing to upload
```
Why should be an error? It is a warning to make the user understand in case of failure the reason of the failure or in case of allowEmptyArchive is true to advise the user that nothing was uploaded.

chrismathis replied:
If allowEmptyArchive is false it should be an error because it's the reason the build failed. In the other case it should be a warning.
For failsAsUpload it would actually be the same.

nfalco79 replied:
Yes it's the reason you can not put this logic in the Uploader but in dronerecorder

nfalco79 commented:
Please remember to use the correct code formatter (preferably only on the portion of the modified code)

chrismathis commented:
Uh, sorry for that - changed
(Also changed the setting for formatting changed lines)

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/DroneRecorder.java:
```diff
@@ -341,45 +339,21 @@ public void perform ( final Run<?, ?> run, final FilePath workspace, final Launc
      */
     private boolean validateStart ( final Run<?, ?> run, final TaskListener listener )
     {
-        if ( this.artifacts == null )
```
Restore this check

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/DroneRecorder.java:
```diff
@@ -485,19 +433,40 @@ public Void invoke ( final File basedir, final VirtualChannel channel ) throws I
-//                        TODO this.artifacts.put ( artifacId, f );
                     }
-
+                    uploadResult.addArtifacts ( uploader.complete () );
```
Put this in a finally block or in case of Uploader V2 you lost the already uploaded artifacts

chrismathis replied:
I think uploader.complete() should not be called when there is an exception in upload()
As of V3 complete is just another step in the upload process.
-> Introduced ` public Map<String, String> getUploadedArtifacts ();` in Uploader

nfalco79 replied:
What about if in V2 upload fails on thrird file? The upload result contains the uploaded files with success?

chrismathis replied:
Yes it will. getUploadedArtifacts will be called in a finally.

nfalco79 replied:
Ok, it will. Than what is the the scope of complete method? If I want implement a new Uploader, what I should perform in the complete method? What does it returns now?

chrismathis replied:
Upload is called for each individual file and does what is necessary for the files. Complete finalized the transaction (maybe does the actual upload (like V3)). It returns void and throws IOExceptions. For V2 it does nothing (I should do an empty implementation in AbstractUploader probably).
getUploadedArtifacts just returns the Map and throws no Exceptions

nfalco79 replied:
> Complete finalized the transaction

It's difficult to think a transaction in a rest client and how it could be treated.
What do you think about
* addArtifact collect the files to send
* upload or performUpload or run execute the real upload and throws exception
* getUploadedArtifacts returns the artifacts uploaded with success
* close clone open connection

It's a little bit clear so V2 and V3 seems have the same behaviour

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/UploaderV3.java:
```diff
@@ -148,61 +142,52 @@ public boolean complete ()
+                            if ( httpResponseErrorMessage != null )
                             {
-                                addErrorMessage ( "Failed to upload: " + response.getStatusLine () );
+                                errorMessage += " " + httpResponseErrorMessage;
```
add a new line to keep the previous behaviour `+= "\n" + httpResponseErrorMessage`

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/UploaderV3.java:
```diff
@@ -211,14 +196,16 @@ private void processUploadResult ( final String string )
         {
             e.printStackTrace ( this.listener.error ( "Failed to parse upload result" ) );
         }
+        return new HashMap<> ();
```
return an empty map `Collections.emptyMap ()`

nfalco79 reviewed (changes requested):
Please provide more commits because the squash make review invalid.
We will provide a single commit when merge using the github feature

chrismathis committed:
Changes form review

*Reenable save actions
*Add javadoc
*Externalize error messages
*Rename Methods of Uploder interface and add getUploadedArtifacts
*Re add accidentally removed parameter check

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/UploaderV2.java:
```diff
@@ -93,7 +91,7 @@ private URI makeUrl ( final String file ) throws URIException, IOException
 
     @Override
-    public void upload ( final File file, final String filename ) throws IOException
+    public void addArtifact ( final File file, final String filename ) throws IOException
```
Collect artifacts to the internal map and than performs the upload in the performUpload method

chrismathis replied:
Thought I'd leave that for another refactoring, but will do that now.

nfalco79 replied:
Which other refactor?

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/UploaderV2.java:
```diff
@@ -162,4 +154,9 @@ public void close ()
     }
 
+    @Override
+    public void performUpload () throws IOException
```
Here should be performed the real upload

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/Uploader.java:
```diff
@@ -13,12 +13,32 @@
 {
-    public void upload ( File file, String filename ) throws IOException;
+    /**
+     * Add a single file for upload
```
The first line of javadoc must always end with a dot.

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/Uploader.java:
```diff
@@ -13,12 +13,32 @@
+    /**
+     * Add a single file for upload
+     *
+     * @param file
```
No description for @param file
`the file to be uploaded.`

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/Uploader.java:
```diff
@@ -13,12 +13,32 @@
+     * Add a single file for upload
+     *
+     * @param file
+     * @param filename
```
No description for @param file, expected something like `the artifact name on the server`

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/Uploader.java:
```diff
@@ -13,12 +13,32 @@
+    public void performUpload () throws IOException;
+
+    /**
+     * @return a map of successfully artifact ids and filenames
```
Miss the description like
```
Returns a map of all artifacts uploaded with success to the server.
@return a map of identifier (assigned by the server) - artifact name upload with success.
```

nfalco79 commented on package-drone/src/main/java/de/dentrassi/pm/jenkins/UploaderResult.java:
```diff
@@ -0,0 +1,88 @@
+    private static final long serialVersionUID = -3089286880912224513L;
+
+    /*
+     * Map containing the id and filename of the successfully uploaded artifacts
```
unless field is public (standatd javadoc generation) move this to the getter/setter or does not appear

chrismathis committed:
Fix Javadoc

chrismathis committed:
V2: do the actual upload in performUpload

nfalco79 commented:
good work if @ctron agree we can merge with a github squash commit.

ctron commented:
Thanks for the great work everyone!

ctron commented:
Should I make a new release?

nfalco79 commented:
Let consider the #19 that looks better instead of `artifactId = filename` and add coverage for V2/V3 implementations
